### PR TITLE
FEATURE: Move TrustedProxiesComponent to Middleware

### PR DESCRIPTION
**Pull request by sorenmalling**
Adjust TrustedProxiesComponent to match PSR-15 middleware implementation

- [x] Adjust/remove component tests

Related #2019

### CODE REVIEW

File: Neos.Flow/Classes/Http/Middleware/TrustedProxiesMiddleware.php [hunk 1]
```diff
@@ -12,15 +14,18 @@
...
 use Neos\Flow\Utility\Ip as IpUtility;
+use Psr\Http\Message\ResponseInterface;
 use Psr\Http\Message\ServerRequestInterface;
-use Neos\Flow\Configuration\Exception\InvalidConfigurationException;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
 
 /**
- * HTTP component that checks request headers against a configured list of trusted proxy IP addresses.
+ * Midd leware that checks request headers against a configured list of trusted proxy IP addresses.
```
Comment (albe) [inline Neos.Flow/Classes/Http/Middleware/TrustedProxiesMiddleware.php:26]:
```suggestion
 * Middleware that checks request headers against a configured list of trusted proxy IP addresses.
```

**Comment by sorenmalling** [comment]
@albe How do we port the tests from TrustedProxiesComponent to test the middleware?

**Comment by albe** [comment]
> @albe How do we port the tests from TrustedProxiesComponent to test the middleware?

Just move the test to `../Middleware/TrustedProxiesMiddlewareTest.php`, then adjust the `callWithRequest()` method to invoke the middleware (`$middleware->process($request, $captureRequestHandler)`), capture the `$request` sent with the `$handler->handle($trustedRequest)` invokation and return that. So `$captureRequestHandler` basically needs to be a spy mock, that records the argument that the `handle()` method is invoked with.

**Comment by sorenmalling** [comment]
> capture the `$request` sent with the `$handler->handle($trustedRequest)` invokation and return that. So `$captureRequestHandler` basically needs to be a spy mock, that records the argument that the `handle()` method is invoked with.

You lost me on that :) how do I capture the request being handled inside the middleware? I can commit what I tried so far if of any help

**Comment by albe** [comment]
I was thinking something like
```php
class SpyRequestHandler {
     protected $handledRequest;
     public function handle(ServerRequest $request) {
         $this->handledRequest = $request;
     }
     public function getHandledRequest() {
         return $this->handledRequest;
    }
}

... 
function callWithRequest($request) {
    ...
    $spyHandler = new SpyRequestHandler();
    $response = $middleware->process($request, $spyHandler);
    return $spyHandler->getHandledRequest();
}
```
Alternatively, a spy is return from the PHPUnit `$this->any()` constraint: https://lyte.id.au/2014/03/01/spying-with-phpunit/
But I guess, that's not less code and is likely a bit less straightforward to grasp?

**Comment by albe** [comment]
Green! 🟢🎉

**Review by albe — APPROVED** [review]

**Comment by albe** [comment]
@bwaidelich I would suggest to merge this now and then rebase the next PR that moves the components to middleware and repeat, wdyt?

**Comment by bwaidelich** [comment]
Makes sense! Just wondering: Will this affect the current documentation?
In any case we should mention here: https://neos-project.slack.com/archives/C04Q0TC15/p1601626445022900

**Comment by albe** [comment]
Yes, it will affect documentation, as that still mentions the component chain. I would suggest to create an own documentation PR for the whole move and link that to the parent issue #2019. I can prepare a first draft for the docs change sometime this week.

**Comment by albe** [comment]
@bwaidelich I'll merge this for now to move forward - if we do find this components->middleware change to not get through completely or create problems, then we can revert before branching off.

**Pull request merged**